### DISCUSSION TITLE
MRS Estimator vs PX4 Estimator and how to incorporate MoCap

### DISCUSSION THREAD

**Discussion by alexkoven** [ctu-mrs/mrs_uav_system#133] [Q&A] [answered]
All,

I have two questions. We are looking to implement the MRS MPC on our research drones. We have already worked with PX4 before. I was wondering what the benefit of the MRS estimator is over the PX4 estimator?

One reason why I ask is that we are currently trying to integrate pose measurements from a MoCap system and this integration was quite easy with the PX4 estimator. We are not quite sure how to best do that with the MRS estimators.  [This paper](https://arxiv.org/pdf/2008.08050.pdf) mentions something about using a disturbance estimator. Do you have more info on this?

Thanks in advance!

**Comment by petrlmat** [comment, marked as answer]
Hi, the main purpose of the MRS estimator is to produce a full-state estimate in the form required by the control pipeline of the MRS system, from various measurements of various states, such as GPS, IMU, rangefinder, barometer, Visual SLAM, LiDAR SLAM, etc. These measurements come in different reference frames at different rates with different delays. The estimator fuses all these sources into a number of state hypotheses, which are continuously monitored for health using input data rate check, gates, filters, etc. so that the most reliable can be selected at any given time, and signal the control pipeline when a fault is detected to perform an emergency landing in time. On the other hand, the PX4 estimator has a more complex model that might produce a more consistent estimate with the low-rate measurement as it can trust the model more.

We currently do not use any MoCap system, so I am not sure about the specifics of its output, but the position could probably be fused as VIO measurements. Fusing orientation and angular rates from the MoCap (if they are even provided) is discouraged, as the delay in these states could easily destabilize the UAV. The disturbance estimator is part of the control algorithm, it is being used by default, and it is tuned to work well under many disturbations such as wind, miscalibrated IMU, imbalanced UAV, an inconsistent tilt of propellers, etc.
